File: server/src/main/java/de/ingrid/igeserver/profiles/mcloud/types/mCloudType.java
package de.ingrid.igeserver.profiles.mcloud.types;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.orientechnologies.orient.core.db.ODatabaseSession;
import com.orientechnologies.orient.core.metadata.schema.OClass;
import com.orientechnologies.orient.core.metadata.schema.OSchema;
import com.orientechnologies.orient.core.metadata.schema.OType;
import de.ingrid.igeserver.api.ApiException;
import de.ingrid.igeserver.db.DBApi;
import de.ingrid.igeserver.documenttypes.AddressWrapperType;
import de.ingrid.igeserver.documenttypes.DocumentType;
import de.ingrid.igeserver.services.DocumentService;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.stereotype.Service;

@Service
public class mCloudType extends DocumentType {

    private static Logger log = LogManager.getLogger(mCloudType.class);

    private static final String TYPE = "mCloudDoc";

    private static final String[] profiles = new String[]{"mcloud"};

    public mCloudType() {

        super(TYPE, profiles);
    }

    @Override
    public void initialize(ODatabaseSession session) {

        OSchema schema = session.getMetadata().getSchema();
        if (!schema.existsClass(TYPE)) {
            log.debug("Create class " + TYPE);
            OClass clazz = schema.createClass(TYPE);
            clazz.createProperty("_id", OType.STRING);
            clazz.createProperty("_parent", OType.STRING);

            OClass addressWrapperClass = session.getClass("AddressWrapper");
            clazz.createProperty("addresses", OType.LINKLIST, addressWrapperClass);
        }

    }

    @Override
    public void handleLinkedFields(JsonNode doc, DBApi dbService) throws ApiException {

        handleLinkedAddresses(doc, dbService);

    }

    @Override
    public void mapLatestDocReference(JsonNode doc, DocumentService docService) {

        handleLatestsAddresses(doc, docService);

    }

    private void handleLatestsAddresses(JsonNode doc, DocumentService docService) {

        JsonNode addresses = doc.path("addresses");

        for (JsonNode address : addresses) {
            String wrapperId = address.path("ref").asText();
            JsonNode wrapper = null;
            try {
                wrapper = docService.getByDocId(wrapperId, AddressWrapperType.TYPE, true);
                ObjectNode latestDocument = docService.getLatestDocument(wrapper);
                ((ObjectNode) address).put("ref", latestDocument);
            } catch (Exception e) {
                log.error(e);
            }
        }

    }

    private void handleLinkedAddresses(JsonNode doc, DBApi dbService) throws ApiException {

        JsonNode addresses = doc.path("addresses");

        for (JsonNode address : addresses) {
            String id = address.path("ref").path("_id").textValue();
            ((ObjectNode) address).put("ref", id);
        }

    }

}

File: server/src/main/java/de/ingrid/igeserver/profiles/mcloud/types/MCloudType.kt
/**
 * ==================================================
 * Copyright (C) 2023-2024 wemove digital solutions GmbH
 * ==================================================
 * Licensed under the EUPL, Version 1.2 or – as soon they will be
 * approved by the European Commission - subsequent versions of the
 * EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */
package de.ingrid.igeserver.profiles.mcloud.types

import com.fasterxml.jackson.databind.JsonNode
import de.ingrid.igeserver.persistence.model.EntityType
import de.ingrid.igeserver.persistence.postgresql.jpa.model.ige.Document
import de.ingrid.igeserver.utils.getBoolean
import de.ingrid.igeserver.utils.getString
import org.apache.logging.log4j.kotlin.logger
import org.springframework.stereotype.Component
import java.net.URLDecoder

@Component
class MCloudType : EntityType() {
    override val className = "mCloudDoc"
    override val profiles = arrayOf("mcloud", "test")

    val log = logger()

    override val jsonSchema = "/mcloud/schemes/mcloud.schema.json"

    override fun getUploads(doc: Document): List<String> {
        return doc.data.get("distributions")
            ?.filter { download -> !(download.getBoolean("link.asLink") ?: false) }
            ?.map { download -> getUploadFile(download) } ?: emptyList()
    }

    private fun getUploadFile(download: JsonNode): String {
        return if (download.get("link").get("uri") != null) {
            URLDecoder.decode(download.getString("link.uri")!!, "utf-8")
        } else {
            download.get("link").get("value").textValue()
        }
    }

    override fun getReferenceIds(doc: Document): List<String> {
        return doc.data.path("addresses").map { address ->
            address.getString("ref")!!
        }
    }
}
